external-crates/move/crates/move-binary-format/src/normalized.rs

```diff
@@ -305,337 +664,400 @@ impl Module {
-        }
-    }
-
-    pub fn into_type_tag(self) -> Option<TypeTag> {
-        use Type::*;
-        Some(if self.is_closed() {
-            match self {
-                Reference(_) | MutableReference(_) => return None,
-                Bool => TypeTag::Bool,
-                U8 => TypeTag::U8,
-                U16 => TypeTag::U16,
-                U32 => TypeTag::U32,
-                U64 => TypeTag::U64,
-                U128 => TypeTag::U128,
-                U256 => TypeTag::U256,
-                Address => TypeTag::Address,
-                Signer => TypeTag::Signer,
-                Vector(t) => TypeTag::Vector(Box::new(
-                    t.into_type_tag()
-                        .expect("Invariant violation: vector type argument contains reference"),
-                )),
-                Struct {
-                    address,
-                    module,
-                    name,
-                    type_arguments,
-                } => TypeTag::Struct(Box::new(StructTag {
-                    address,
-                    module,
-                    name,
-                    type_params: type_arguments
-                        .into_iter()
-                        .map(|t| {
-                            t.into_type_tag().expect(
-                                "Invariant violation: struct type argument contains reference",
-                            )
-                        })
-                        .collect(),
-                })),
-                TypeParameter(_) => unreachable!(),
-            }
-        } else {
-            return None;
+    /// How often are signatures used in this module?
+    pub fn signature_usage(&self) -> usize {
+        self.tables.signatures.iter().fold(0, |acc, sig_rc| {
+            // subtract one for the reference in the signature table
+            acc.saturating_add(Rc::strong_count(sig_rc).saturating_sub(1))
         })
     }
 
-    pub fn into_struct_tag(self) -> Option<StructTag> {
-        match self.into_type_tag()? {
-            TypeTag::Struct(s) => Some(*s),
-            _ => None,
-        }
-    }
-
-    pub fn subst(&self, type_args: &[Type]) -> Self {
-        use Type::*;
-        match self {
-            Bool | U8 | U16 | U32 | U64 | U128 | U256 | Address | Signer => self.clone(),
-            Reference(ty) => Reference(Box::new(ty.subst(type_args))),
-            MutableReference(ty) => MutableReference(Box::new(ty.subst(type_args))),
-            Vector(t) => Vector(Box::new(t.subst(type_args))),
-            Struct {
-                address,
-                module,
-                name,
-                type_arguments,
-            } => Struct {
-                address: *address,
-                module: module.clone(),
-                name: name.clone(),
-                type_arguments: type_arguments
+    /// Panics if called with `include_code` set to `false`.
+    pub fn equals(&self, other: &Self) -> bool {
```

Comment:
... But the efficiency flag `includes_code` comes at a cost, where we want to be very careful of not checking equality between functions when the bodies are elided. Alternatively, we could make two different wrapper types around Module and Function. Instead of that, I made this panic and removed the `Eq` trait to avoid accidental usage.